File: quiz_engine/src/exceptions/QuizExceptions.ts
export class QuizException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'QuizException';
  }
}

export class QuestionNotFoundError extends QuizException {
  constructor(id: number) {
    super(`Question with id ${id} not found`);
    this.name = 'QuestionNotFoundError';
  }
}

export class SessionNotFoundError extends QuizException {
  constructor(sessionId: string) {
    super(`Quiz session '${sessionId}' not found`);
    this.name = 'SessionNotFoundError';
  }
}

export class InsufficientQuestionsError extends QuizException {
  constructor(requested: number, available: number) {
    super(`Requested ${requested} questions but only ${available} available`);
    this.name = 'InsufficientQuestionsError';
  }
}

export class InvalidAnswerError extends QuizException {
  constructor(answer: string) {
    super(`Invalid answer: '${answer}'. Must be A, B, C, D, or E`);
    this.name = 'InvalidAnswerError';
  }
}

export class ParseError extends QuizException {
  constructor(message: string) {
    super(`Parse error: ${message}`);
    this.name = 'ParseError';
  }
}

export class DatabaseError extends QuizException {
  constructor(message: string) {
    super(`Database error: ${message}`);
    this.name = 'DatabaseError';
  }
}

File: quiz-engine/quiz-engine-dart/lib/src/exceptions/quiz_exceptions.dart
/// Custom exceptions for the quiz engine.
library;

/// Thrown when no questions are found in the database.
class NoQuestionsException implements Exception {
  final String message;
  const NoQuestionsException([this.message = 'No questions found in database.']);

  @override
  String toString() => 'NoQuestionsException: $message';
}

/// Thrown when the requested question count exceeds available questions.
class InsufficientQuestionsException implements Exception {
  final int requested;
  final int available;

  const InsufficientQuestionsException({
    required this.requested,
    required this.available,
  });

  @override
  String toString() =>
      'InsufficientQuestionsException: Requested $requested questions but only $available available.';
}

/// Thrown when a quiz session is not found.
class SessionNotFoundException implements Exception {
  final String sessionId;
  const SessionNotFoundException(this.sessionId);

  @override
  String toString() => 'SessionNotFoundException: Session "$sessionId" not found.';
}

/// Thrown when a file cannot be imported.
class ImportException implements Exception {
  final String path;
  final String reason;

  const ImportException({required this.path, required this.reason});

  @override
  String toString() => 'ImportException: Cannot import "$path": $reason';
}

/// Thrown when a database operation fails.
class DatabaseException implements Exception {
  final String message;
  const DatabaseException(this.message);

  @override
  String toString() => 'DatabaseException: $message';
}
